compiler/src/dotty/tools/dotc/util/Signatures.scala

```diff
@@ -41,109 +43,238 @@ object Signatures {
+    span: Span,
+    params: List[tpd.Tree],
+    fun: tpd.Tree
+  )(using Context): (Int, Int, List[Signature]) =
+    val (alternativeIndex, alternatives) = fun.tpe match
       case err: ErrorType =>
-        val (alternativeIndex, alternatives) = alternativesFromError(err, params)
+        val (alternativeIndex, alternatives) = alternativesFromError(err, params) match
+          // if we have no alternatives from error, we have to fallback to function denotation
+          // Check `partialyFailedCurriedFunctions` test for example
+          case (_, Nil) =>
+            val denot = fun.denot
+            if denot.exists then
+              (0, List(denot.asSingleDenotation))
+            else
+              (0, Nil)
+          case other => other
         (alternativeIndex, alternatives)
-
       case _ =>
         val funSymbol = fun.symbol
         val alternatives = funSymbol.owner.info.member(funSymbol.name).alternatives
         val alternativeIndex = alternatives.map(_.symbol).indexOf(funSymbol) max 0
         (alternativeIndex, alternatives)
-    }
 
-    (paramIndex, alternativeIndex, alternatives)
+    val curriedArguments = countParams(fun, alternatives(alternativeIndex))
+    val paramIndex = params.indexWhere(_.span.contains(span)) match {
+      case -1 => (params.length - 1 max 0) + curriedArguments
+      case n => n + curriedArguments
+    }
+    val alternativeSignatures = alternatives.flatMap(toApplySignature)
+    (paramIndex, alternativeIndex, alternativeSignatures)
 
+  /**
+   * Extracts call informatioin for function in unapply context.
+   *
+   * @param span The position of the cursor
+   * @param params Current function parameters
+   * @param fun Unapply function tree
+   *
+   * @return A triple containing the index of the parameter being edited, the index of the function
+   *         being called, the list of overloads of this function).
+   */
   private def unapplyCallInfo(
     span: Span,
     fun: tpd.Tree,
     patterns: List[tpd.Tree]
-  )(using Context): (Int, Int, List[SingleDenotation]) =
-    val patternPosition = patterns.indexWhere(_.span.contains(span))
-    val activeParameter = extractParamTypess(fun.tpe.finalResultType.widen).headOption.map { params =>
-      (patternPosition, patterns.length) match
-        case (-1, 0) => 0 // there are no patterns yet so it must be first one
-        case (-1, pos) => -1 // there are patterns, we must be outside range so we set no active parameter
-        case _ => (params.size - 1) min patternPosition max 0 // handle unapplySeq to always highlight Seq[A] on elements
-    }.getOrElse(-1)
+  )(using Context): (Int, Int, List[Signature]) =
+    val resultType = unapplyMethodResult(fun)
+    val isUnapplySeq = fun.denot.name == core.Names.termName("unapplySeq")
+    val denot = fun.denot
+
+    def extractParamTypess(resultType: Type): List[List[Type]] =
+      resultType match
+        // Reference to a type which is not a type class
```

Comment:
It's probably better to show code examples here. Can do without an explanation even.